packages/server/utils/__tests__/rateLimiters/InMemoryRateLimiter.test.ts

```diff
@@ -0,0 +1,182 @@
+      // Given an InMemoryRateLimiter
+      const inMemoryRateLimiter = subject()
+      const userId = '123'
+      const fieldName = 'FieldName'
+
+      // When the same user requests the same field twice within the same minute
+      inMemoryRateLimiter.log(userId, fieldName, true)
+      jest.advanceTimersByTime(ms('30s'))
+      const secondResult = inMemoryRateLimiter.log(userId, fieldName, true)
+
+      // Then it returns 2 calls last minute, 2 calls last hour
+      expect(secondResult).toEqual({lastMinute: 2, lastHour: 2})
+    })
+
+    it('returns 1 call last minute, 2 calls last hour when two calls occur a minute apart', () => {
+      // Given an InMemoryRateLimiter
+      const inMemoryRateLimiter = subject()
+      const userId = '123'
+      const fieldName = 'FieldName'
+
+      // When the same user requests the same field one minute apart
+      inMemoryRateLimiter.log(userId, fieldName, true)
+      jest.advanceTimersByTime(ms('1m'))
+      const secondResult = inMemoryRateLimiter.log(userId, fieldName, true)
+
+      // Then it returns 1 call last minute, 2 calls last hour
+      expect(secondResult).toEqual({lastMinute: 1, lastHour: 2})
+    })
+
+    it('returns 1 call last minute, 1 calls last hour when two calls occur an hour apart', () => {
+      // Given an InMemoryRateLimiter
+      const inMemoryRateLimiter = subject()
+      const userId = '123'
+      const fieldName = 'FieldName'
+
+      // When the same user requests the same field one hour apart
+      inMemoryRateLimiter.log(userId, fieldName, true)
+      jest.advanceTimersByTime(ms('1h'))
+      const secondResult = inMemoryRateLimiter.log(userId, fieldName, true)
+
+      // Then it returns 1 call last hour, 1 call last minute
+      expect(secondResult).toEqual({lastMinute: 1, lastHour: 1})
+    })
+
+    it('does not reuse the same rate limit for different users', () => {
+      // Given an InMemoryRateLimiter
+      const inMemoryRateLimiter = subject()
+      const firstUserId = '123'
+      const secondUserId = '456'
+      const fieldName = 'FieldName'
+
+      // When a different user requests the same field 1 minute apart
+      inMemoryRateLimiter.log(firstUserId, fieldName, true)
+      jest.advanceTimersByTime(ms('1m'))
+      const secondUserFirstCallResult = inMemoryRateLimiter.log(secondUserId, fieldName, true)
+
+      // Then it returns 1 call last minute, 1 call lastHour
+      expect(secondUserFirstCallResult).toEqual({lastMinute: 1, lastHour: 1})
+    })
+  })
+
+  describe('gc', () => {
+    it('cleans up users who have not called the service in the last hour', () => {
+      // Given an InMemoryRateLimiter with scheduled GC enabled
+      const inMemoryRateLimiter = subject({scheduleGc: true})
+      const userId = '123'
+      const fieldName = 'FieldName'
+
+      // When a user requested a field
+      inMemoryRateLimiter.log(userId, fieldName, false)
+      // When two hours elapse (triggering GC twice)
```

Comment:
I like given/when/then style testing: https://martinfowler.com/bliki/GivenWhenThen.html. I think it helps make tests, especially more complicated tests like this one, more understandable.

Reply:
+1 I like structuring the tests in these 3 sections (I know this as arrange, act, assert / AAA), but I'm not fond of all those verbose comments repeating the test description or the code itself. I would prefer some middle ground only stating the non-obvious, otherwise I find these distract me from the actual code. If it's mostly about the structure, I think
```ts
      // Given
      const inMemoryRateLimiter = subject({scheduleGc: true})
      const userId = '123'
      const fieldName = 'FieldName'

      // When two hours elapse between requests, triggering GC twice
      inMemoryRateLimiter.log(userId, fieldName, false)
      jest.advanceTimersByTime(ms('1h'))
      jest.advanceTimersByTime(ms('1h'))
      const secondResult = inMemoryRateLimiter.log(userId, fieldName, false)

      // Then
      // We're relying on the behavior that you receive 1 lastHour the first time the user calls without
      // extended logging for this assertion. If the user was not GC'ed we've get lastMinute: 1, lastHour: undefined
      expect(secondResult).toEqual({lastMinute: 1, lastHour: 1})
```
would be verbose enough.

Reply:
Sounds good - I removed some verbosity.